### ISSUE TITLE
Descriptor __get__ fails to match Coroutine methods as (...) -> Awaitable[R]

### ISSUE THREAD

**Issue by maxfischer2781** [microsoft/pyright#7533]
**Describe the bug**
When a descriptor captures an `async def` method, the descriptor ``__get__`` should be able to destructure it as `Callable[..., Awaitable[R]]` to extract the return type. However, PyRight does not match this case and consequently fails to understand such descriptors.

> Cannot access member ...
> Failed to call method "__get__" for descriptor class ...

Notably, MyPy accepts such constructs and infers everything as expected.

**Code or Screenshots**

This is a MRE derived from an [async cache decorator](https://github.com/maxfischer2781/asyncstdlib/blob/8402004ab30bf327314b369e2811d89f3955f250/asyncstdlib/_lrucache.pyi#L33-L42). Notably, PyRight does *not* reject decoration itself, which should mean the coroutine function matches against the TypeVar bound `Callable[..., Awaitable[Any]]`. The issue also occurs if `__get__` uses `Callable[..., Awaitable[Any]]`, so it seems not caused by matching the TypeVar `R` (but the example is pointless without it).

```python
from typing import Any, Awaitable, Callable, Coroutine, Generic, TypeVar, overload

AC = TypeVar("AC", bound=Callable[..., Awaitable[Any]])
R = TypeVar("R")

# descriptor
class ADescr(Generic[AC]):
    def __init__(self, callable: AC): ...

    @overload
    def __get__(
        self: ADescr[AC], instance: None, owner: type | None = ...
    ) -> ADescr[AC]: ...
    @overload
    def __get__(
        self: ADescr[Callable[..., Awaitable[R]]], instance: object, owner: type | None = ...,
    ) -> Callable[..., Awaitable[R]]: ...

    def __get__(self, instance: object | None, owner: type | None = None) -> Any: ...

# test case
class ATest:
    @ADescr
    async def foo(self, a: int) -> int: ...

    async def access(self):
        return await self.foo(12)
```
For the last line, pyright reports the following error:
```
29:27 - error: Cannot access member "foo" for type "ATest*"
    Failed to call method "__get__" for descriptor class "ADescr[(self: Self@ATest, a: int) -> Coroutine[Any, Any, int]]" (reportAttributeAccessIssue)
```

**VS Code extension or command-line**
Tested both via PyLance in VSCode as well as the Python command line wrapper for PyRight 1.1.355.

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The class `ADesc` is parameterized by an invariant type variable `AC`. That means the type arguments must match exactly for the types to be compatible. The post-decorated type of `foo` is `ADesc[(self, a: int) -> Coroutine[Any, Any, int]`. The annotated type of the `self` parameter for the second overload of the `__get__` method is `ADesc[Callable[..., Awaitable[R]]`. While `Coroutine[Any, Any, int]` is a subtype of `Awaitable[R]`, they are not the same, so the compatibility check fails due to invariance.

You can fix the problem in the code by making `AC` covariant.

```python
AC = TypeVar("AC", bound=Callable[..., Awaitable[Any]], covariant=True)
```

Or if you are using Python 3.12, you can switch to the new type parameter syntax introduced in PEP 695 and no longer worry about variance because the type checker will determine the correct variance based on the class definition.

**Comment by maxfischer2781** [comment]
Thanks, that makes sense and `covariant=True` indeed fixes the issue.

From a usability perspective, should PyRight accept applying the decorator? This too must match `AC` against the `(...) -> Coroutine` but no error is shown there. Hence, me mistakenly thinking that the type variable was correct.
